### ISSUE TITLE
Grab one sample from pipeline for shaping and reset pipeline

### ISSUE THREAD

**Issue by jramapuram** [NVIDIA/DALI#5275]
### Describe the question.

Is it possible to grab a single sample of `(imgs, labels)` from the pipeline and reset it? This is mainly for shape checking.

I see in the docs that:
- `pipe.reset()` is only usable at the end of iteration?
- If I iterate the pipeline for one step the pipeline is missing 1 minibatch for training. 

Why?
- Want to grab shaping / labels for model creation, etc.

### Check for duplicates

- [X] I have searched the [open bugs/issues](https://github.com/NVIDIA/DALI/issues) and have found no duplicates for this bug report

**Comment by JanuszL** [comment]
Hi @jramapuram,

Thank you for reaching out.
In your case, you can either recreate the pipeline to fully reset it, or keep the output of the first iteration and reuse it in further training.
Also in many cases, the pipeline definition determines the output shape - it is a common pattern to use resize/crop operation to have a fixed shape at the data processing output.
Can you also describe in more detail your use case and why you need to run the processing to determine the shape?

**Comment by jramapuram** [comment]
Heya @JanuszL -- hope you are well!

> Also in many cases, the pipeline definition determines the output shape - it is a common pattern to use resize/crop operation to have a fixed shape at the data processing output.
Good point. This might be the smarter option actually.

> Can you also describe in more detail your use case and why you need to run the processing to determine the shape?
It is mainly to handle pipelines that return augmentation multiplicity > 1 (eg: DINO and other self-supervised algorithms use 2 images of resolution1 and 10x images of resolution2 in a minibatch). But you are right. This should be determinable apriori.

**Comment by jramapuram** [comment]
@JanuszL : I tried deleting the pipeline and re-creating it, however there seems to be a weird oddity.

| GPU Memory (%)  | `del dataloader`  |
| ------------- | ------------- |
| 68.76%  | N/A  |
| 77.1%  | yes  |

**Comment by jramapuram** [comment]
@JanuszL : I tried deleting the pipeline and re-creating it, however there seems to be a weird oddity with GPU memory over the entire training run 😬 

| GPU Memory (%)  | `del dataloader`  |
| ------------- | ------------- |
| 68.76%  | N/A (baseline single pipeline creation)  |
| 77.1%  | yes  |
| 76.9%  | yes  |

These are on an A100 80G x8 instance.

**Comment by JanuszL** [comment]
Hi @jramapuram,

If you want to release DALI memory that is kept in the pool, please check [this API](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/advanced_topics_performance_tuning.html#freeing-memory-pools).

**Comment by jramapuram** [comment]
Still not working as expected @JanuszL -- memory at 77%. Any thoughts?

```python
    class DataLoader:
        self.train_loader, self.test_loader, self.valid_loader = self._recreate_dataloaders()

        # ...

        self._delete_loaders()
        self.train_loader, self.test_loader, self.valid_loader = self._recreate_dataloaders()

    def _delete_loaders(self):
        """Delete loaders if they exist."""
        if self.train_loader is not None:
            del self.train_loader

        if self.test_loader is not None:
            del self.test_loader

        if self.valid_loader is not None:
            del self.valid_loader

        # Release memory in pool
        ReleaseUnusedMemory()

        # Force garbage collection
        gc.collect()

**Comment by JanuszL** [comment]
@jramapuram,

Can you provide a standalone repro for the case you see? Now I can only guess what may be going on, and I can just miss details that differentiate your code from the case I have in mind.. It can even be a toy example using data from https://github.com/NVIDIA/DALI_extra.
